src/pydicom/pixels/decoders/base.py

```diff
@@ -1177,17 +1189,33 @@ def _as_array_native(runner: DecodeRunner, index: int | None) -> "np.ndarray":
                 arr = arr.view(dtype)[:length_bytes]
         else:
             if index is not None:
-                start_offset = file_offset + index * length_bytes
+                start_offset = floor(file_offset + index * length_bytes)
                 if (start_offset + length_bytes) > file_offset + length_source:
                     raise ValueError(
                         f"There is insufficient pixel data to contain {index + 1} frames"
                     )
 
-                frame = runner.get_data(src, start_offset, length_bytes)
+                if runner.bits_allocated == 1:
+                    if length_pixels % 8 != 0:
+                        bit_offset_start = (index * length_pixels) % 8
+                        bit_offset_end = -(-((index + 1) * length_pixels) % 8)
+                        if bit_offset_end == 0:
+                            bit_offset_end = None
+
+                frame = runner.get_data(src, start_offset, ceil(length_bytes))
                 arr = np.frombuffer(frame, dtype=dtype)
             else:
                 length_bytes *= runner.number_of_frames
-                buffer = runner.get_data(src, file_offset, length_bytes)
+
+                if runner.bits_allocated == 1:
+                    if length_pixels % 8 != 0:
+                        bit_offset_end = -(
+                            -(runner.number_of_frames * length_pixels) % 8
+                        )
+                        if bit_offset_end == 0:
+                            bit_offset_end = None
```

Comment:
Probably not needed?

Reply:
Addressed in [896c6b1](https://github.com/pydicom/pydicom/pull/2140/commits/896c6b1fc48f431d63e60d944aedd6d53a947293)